### PR TITLE
IIR refactor II

### PR DESCRIPTION
**Pull request by cosunae**
Technical Description
==================

* Move all the IIRNode to std::unique_ptr
* Fix the IIRNode (in particular the keeping the parent_ pointers of a node always in good sync)
* Provide a checkTreeConsistency() function

Notice the TestIIRNode is implemented in #9

### CODE REVIEW

File: src/dawn/IIR/BlockStatements.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  BlockStatements blockStatements;
+
+  for(const auto& stmt : blockStatements_) {
+    blockStatements.insert(stmt->clone());
+  }
+  return blockStatements;
+}
+
+void BlockStatements::insert(std::unique_ptr<StatementAccessesPair>&& stmt) {
+  std::cout << "inser " << blockStatements_.size() << std::endl;
```
Comment (twicki) [inline src/dawn/IIR/BlockStatements.cpp:44]: typo

File: src/dawn/IIR/DoMethod.h [hunk 2]
```diff
@@ -34,25 +34,29 @@ class StatementAccessesPair;
...
   Interval interval_;
-  const long unsigned int id_;
+  long unsigned int id_;
 
   std::shared_ptr<DependencyGraphAccesses> dependencyGraph_;
-  //  std::vector<std::shared_ptr<StatementAccessesPair>> statementAccessesPairs_;
+  using base_type = IIRNode<Stage, DoMethod, StatementAccessesPair>;
 
 public:
+  static constexpr const char* name = "DoMethod";
```
Comment (twicki) [inline src/dawn/IIR/DoMethod.h:43]: why do we want the name?

File: src/dawn/IIR/IIR.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+namespace iir {
+
+namespace impl {
+template <typename T>
+using StencilSmartptr = std::unique_ptr<T, std::default_delete<T>>;
+}
+
+/// @brief A Stencil is represented by a collection of MultiStages
+/// @ingroup optimizer
+class IIR : public IIRNode<void, IIR, Stencil /*, impl::StencilSmartptr*/> {
```
Comment (twicki) [inline src/dawn/IIR/IIR.h:31]: container?

File: src/dawn/IIR/IIRNode.h [hunk 4]
```diff
@@ -39,20 +46,24 @@ bool ptrEqual(const std::weak_ptr<P>& f, const std::weak_ptr<P>& s) {
...
 class IIRNode {
 
 protected:
   IIRNode() = default;
-  IIRNode(const IIRNode&) = default;
   IIRNode(IIRNode&&) = default;
 
-  std::weak_ptr<Parent> parent_;
-  /// List of Do-Methods of this stage
+  std::unique_ptr<Parent> const* parent_ = nullptr;
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h:56]: const x vs x const

File: src/dawn/IIR/MultiStage.cpp [hunk 5]
```diff
@@ -15,6 +15,7 @@
 #include "dawn/IIR/MultiStage.h"
 #include "dawn/IIR/Accesses.h"
 #include "dawn/IIR/DependencyGraphAccesses.h"
+// TODO remove as much as possible includes from Optimizer
```
Comment (twicki) [inline src/dawn/IIR/MultiStage.cpp:18]: still open?

File: src/dawn/IIR/MultiStage.h [hunk 6]
```diff
@@ -54,26 +55,30 @@ class MultiStage : public IIRNode<Stencil, MultiStage, Stage, std::shared_ptr, i
...
+  static constexpr const char* name = "MultiStage";
+
   using StageSmartPtr_t = child_smartptr_t<Stage>;
   using ChildrenIterator = std::vector<child_smartptr_t<Stage>>::iterator;
 
   /// @name Constructors and Assignment
   /// @{
   MultiStage(StencilInstantiation& stencilInstantiation, LoopOrderKind loopOrder);
-  MultiStage(const MultiStage&) = default;
+  //  MultiStage(const MultiStage&) = default;
```
Comment (twicki) [inline src/dawn/IIR/MultiStage.h:66]: rm

File: src/dawn/IIR/IIRNode.h [hunk 7]
```diff
@@ -62,153 +73,389 @@ class IIRNode {
...
+    for(const auto& child : getChildren()) {
+      if(!child->parentIsSet()) {
+        return false;
+      }
+      const auto& parent = child->getParent();
+
+      if(parent.get() != this) {
+        return false;
+      }
+      result = result & child->checkTreeConsistency();
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h:313]: do we want early exit?

File: src/dawn/IIR/IIRNode.h [hunk 8]
```diff
@@ -62,153 +73,389 @@ class IIRNode {
...
+  void setChildrenParent(const std::unique_ptr<TNode>& p,
+                         typename std::enable_if<is_child_void<TChildSmartPtr>::value>::type* = 0) {
+  }
+
+  template <typename TParent, typename TChild>
+  void setChildParent(
+      const std::unique_ptr<TChild>& child,
+      typename std::enable_if<!std::is_void<TParent>::value &&
+                              !is_child_void<std::unique_ptr<TChild>>::value>::type* = 0) {
+    static_assert(std::is_same<TParent, Parent>::value,
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h:197]: use the macro?

File: src/dawn/IIR/IIRNode.h [hunk 9]
```diff
@@ -62,153 +73,389 @@ class IIRNode {
...
+
+private:
+  template <typename TChild>
+  void cloneChildrenImpl(const IIRNode& other,
+                         typename std::enable_if<std::is_void<TChild>::value>::type* = 0) {}
+
+  template <typename TChild>
+  void cloneChildrenImpl(const IIRNode& other,
+                         typename std::enable_if<!std::is_void<TChild>::value>::type* = 0) {
+    static_assert(std::is_same<TChild, Child>::value,
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h:271]: macro?

File: src/dawn/IIR/IIRNode.h [hunk 10]
```diff
@@ -62,153 +73,389 @@ class IIRNode {
...
-    insertChildrenImpl<Iterator, Parent>(pos, first, last);
+  template <typename TChild, typename TParent>
+  ChildIterator insertChildImpl(ChildIterator pos, TChild&& child,
+                                typename std::enable_if<!std::is_void<TParent>::value>::type* = 0) {
+
+    PROTECT_TEMPLATE(TParent, Parent)
+    auto it = children_.insert(pos, std::forward<TChild>(child));
+
+    if(!children_.empty()) {
+      const auto& lastChild = children_.back();
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h:340]: rm this

File: src/dawn/IIR/IIRNode.h [hunk 11]
```diff
@@ -62,153 +73,389 @@ class IIRNode {
...
+  ChildIterator
+  insertChildrenImpl(ChildIterator pos, Iterator first, Iterator last,
+                     typename std::enable_if<!std::is_void<TParent>::value>::type* = 0) {
+    PROTECT_TEMPLATE(TParent, Parent)
+
+    auto newfirst = children_.insert(pos, first, last);
+
+    // setting recursively the children parents is perform on all the siblings of a child, therefore
+    // here we only need to do it on one child
+    if(!children_.empty()) {
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h:356]: again, conditional can be removed

File: src/dawn/IIR/Stage.cpp [hunk 12]
```diff
@@ -26,13 +26,33 @@
...
 
+Stage::Stage(StencilInstantiation& context, MultiStage* multiStage, int StageID)
+    : stencilInstantiation_(context), multiStage_(multiStage), StageID_(StageID),
+      extents_{0, 0, 0, 0, 0, 0} {}
+
 Stage::Stage(StencilInstantiation& context, MultiStage* multiStage, int StageID,
              const Interval& interval)
     : stencilInstantiation_(context), multiStage_(multiStage), StageID_(StageID),
       extents_{0, 0, 0, 0, 0, 0} {
+  // TODO reconsider whether we want to insert an interval
```
Comment (twicki) [inline src/dawn/IIR/Stage.cpp:37]: why should we have the intervals at any other level?

File: src/dawn/IIR/Stage.h [hunk 13]
```diff
@@ -40,11 +40,10 @@ class MultiStage;
...
 ///
 /// @ingroup optimizer
-namespace impl {
-template <typename T>
-using DoMethodSmartptr = std::unique_ptr<T, std::default_delete<T>>;
-}
-class Stage : public IIRNode<MultiStage, Stage, DoMethod, std::shared_ptr> {
+class Stage : public IIRNode<MultiStage, Stage, DoMethod> {
+
+  using base_type = IIRNode<MultiStage, Stage, DoMethod>;
```
Comment (twicki) [inline src/dawn/IIR/Stage.h:45]: why do we need the base_type defn?

File: src/dawn/IIR/StencilFunctionInstantiation.cpp [hunk 14]
```diff
@@ -39,6 +39,35 @@ StencilFunctionInstantiation::StencilFunctionInstantiation(
   DAWN_ASSERT(function);
 }
 
+StencilFunctionInstantiation StencilFunctionInstantiation::clone() const {
+  StencilFunctionInstantiation stencilFun(stencilInstantiation_, expr_, function_, ast_, interval_,
+                                          isNested_);
+
+  stencilFun.hasReturn_ = hasReturn_;
```
Comment (twicki) [inline src/dawn/IIR/StencilFunctionInstantiation.cpp:46]: we want a derived struct that you could copy. I assume this is coming in PR IV?

File: src/dawn/IIR/StencilInstantiation.cpp [hunk 15]
```diff
@@ -1294,20 +1311,23 @@ struct PrintDescLine {
 
 } // anonymous namespace
 
-void StencilInstantiation::checkConsistency() const {
-  for(const auto& stencil : stencils_) {
-    stencil->checkConsistency();
+bool StencilInstantiation::checkTreeConsistency() const {
+  bool result = true;
+  for(const auto& stencil : getStencils()) {
```
Comment (twicki) [inline src/dawn/IIR/StencilInstantiation.cpp:1316]: why not just checkConsistency on the IIR?

File: src/dawn/Optimizer/PassManager.cpp [hunk 16]
```diff
@@ -51,6 +51,8 @@ bool PassManager::runPassOnStecilInstantiation(
     return false;
   }
 
+  DAWN_ASSERT(instantiation->getIIR()->checkTreeConsistency());
```
Comment (twicki) [inline src/dawn/Optimizer/PassManager.cpp:54]: assert with message what pass failed it?

File: src/dawn/IIR/BlockStatements.cpp (hunk 1 above)
Comment (cosunae) [reply src/dawn/IIR/BlockStatements.cpp:44]: this was for debugging, it was removed

File: src/dawn/IIR/DoMethod.h (hunk 2 above)
Comment (cosunae) [reply src/dawn/IIR/DoMethod.h:43]: For debugging, if you have recursive methods for examples that traverse the whole IIR tree and check for consistency of pointer, it is useful to know in which level you are. 
I think it is not used however... at the moment

File: src/dawn/IIR/IIR.h (hunk 3 above)
Comment (cosunae) [reply src/dawn/IIR/IIR.h:31]: fixed

File: src/dawn/IIR/IIRNode.h (hunk 4 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.h:56]: fixed

File: src/dawn/IIR/MultiStage.cpp (hunk 5 above)
Comment (cosunae) [reply src/dawn/IIR/MultiStage.cpp:18]: I did a massive clean up , but still IIR depends on a few algorithms from the Optimizer subproject. 
We should do the cleaning later, but I will remove the comment for now

File: src/dawn/IIR/MultiStage.h (hunk 6 above)
Comment (cosunae) [reply src/dawn/IIR/MultiStage.h:66]: done

File: src/dawn/IIR/IIRNode.h (hunk 7 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.h:313]: done

File: src/dawn/IIR/IIRNode.h (hunk 8 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.h:197]: which macro? This is a static (i.e. not DAWN_ASSERT), do we have anything for this?

File: src/dawn/IIR/IIRNode.h (hunk 9 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.h:271]: same

File: src/dawn/IIR/IIRNode.h (hunk 10 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.h:340]: done

File: src/dawn/IIR/IIRNode.h (hunk 11 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.h:356]: done

File: src/dawn/IIR/Stage.cpp (hunk 12 above)
Comment (cosunae) [reply src/dawn/IIR/Stage.cpp:37]: yeah, the comment was precisely that intervals do not belong to stage, so it is weird that to construct a stage we need to pass an interval, The problem is that this stage ctr is also constructing a do method, (just one), which sounds weird to me. But ok, I removed the TODO and leave it like that

File: src/dawn/IIR/Stage.h (hunk 13 above)
Comment (cosunae) [reply src/dawn/IIR/Stage.h:45]: removed

File: src/dawn/IIR/StencilFunctionInstantiation.cpp (hunk 14 above)
Comment (cosunae) [reply src/dawn/IIR/StencilFunctionInstantiation.cpp:46]: indeed, this is not done yet, not even in PR V

File: src/dawn/IIR/StencilInstantiation.cpp (hunk 15 above)
Comment (cosunae) [reply src/dawn/IIR/StencilInstantiation.cpp:1316]: done

File: src/dawn/Optimizer/PassManager.cpp (hunk 16 above)
Comment (cosunae) [reply src/dawn/Optimizer/PassManager.cpp:54]: done

File: src/dawn/IIR/DoMethod.h (hunk 2 above)
Comment (twicki) [reply src/dawn/IIR/DoMethod.h:43]: we can keep it in if you think it adds value (maybe for a std::out dump?)

File: src/dawn/IIR/IIRNode.h (hunk 8 above)
Comment (twicki) [reply src/dawn/IIR/IIRNode.h:197]: `PROTECT_TEMPLATE`

File: src/dawn/IIR/IIRNode.h (hunk 9 above)
Comment (twicki) [reply src/dawn/IIR/IIRNode.h:271]: `PROTECT_TEMPLATE` from line 29

**Pull request closed without merging**